### PR TITLE
[sessiond] New monitor credit algorithm 

### PR DESCRIPTION
**Pull request by uri200**
Signed-off-by: Oriol Batalla <obatalla@fb.com>

## Summary

The current monitor algorithm stops reporting as soon as it receives a 0 grant. Since we report the update request at 80% of quota, that means that 20% of the quota will not be utilized when 0 is received. This PR changes how the GX (monitors) that and allows the monitor to finish any quota left when receiving a 0 grant

The changes are
- Addition of `received_granted_units` which will record the last grant received. With that we will cal be able to calculate `should_disable_monitor`

- Change on the logic of monitor reporting: when 0 grant is received, we will only report when quota is totally exhausted (we will not report at 80% since we already know that there is no more quota left). Also quen quota is exhausted we will mark the monitor as deleted. But the delete action will not be performed until the next iteration when we get some more traffic for that monitor. The reason for that is we want to report the last chucnk of data, so we can not delete the monitor when we detect it. We could delete the monitor when doing `update_session_credits_and_rules` or `update_sessions`, but to me it made more sense to force it until the next iteration and keep the deletion at `SessionState`

- Change of some related unit test and add new ones. Note that the changes on the unit test forces that last iteration to see if the monitor is deleted and doesn't trigger any update.

## Test Plan

Tested different version on terra VM. 
make test on AGW
cwag_integ_test

Yet, it is very difficult to assure all the edge cases. So we will have to monitor for possible bugs

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by themarwhal** [comment]
So is the goal to not treat a 0 grant as a DISABLE, but as a 0 grant? Even if we receive a grant that includes 0, let the monitor consume all remaining quota before disabling the monitor? Kind of like Gy's FUA, but instead of terminating/redirect, we stop reporting?

**Comment by uri200** [comment]
Yes, that is the idea. Gx doesn't have FAU, so I understand 0 is kind of its final action. When we get 0 that means that no more quota is available from the PCRF, but we don't terminate because we still have some quota left (20% of last grant approximately).

There is an specific AVP to stop reporting, no matter if we still have quota left that will be implemented soon.

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 1]
```diff
@@ -1373,6 +1402,13 @@ void SessionState::set_monitor(
   monitor_map_[key] = std::make_unique<Monitor>(monitor);
 }
 
+bool SessionState::should_delete_monitor(SessionCredit& credit){
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionState.cpp]: Can we attach this function to the Monitor struct? Also we can just do `return credit.current_grant_contains_zero() && credit.is_quota_exhausted(1);` here right?

**Review by themarwhal — APPROVED** [review]

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 1 above)
Comment (uri200) [reply lte/gateway/c/session_manager/SessionState.cpp]: i forgot about monitor structure. Good catch.
BTW, i put the logic in .h file. Not sure if that is a good practice or not🤫

**Comment by uri200** [comment]
just rebased

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 1 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/SessionState.cpp]: Yeah we can move that out to its own CPP/H. I think it's probably not the best practice, as if the header is included in two places it might cause problems. (not 100% sure). :D
